blaze-core/src/main/scala/org/http4s/blazecore/websocket/Serializer.scala

```diff
@@ -0,0 +1,150 @@
+package org.http4s.blazecore.websocket
+
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.{AtomicInteger, AtomicReference}
+
+import org.http4s.blaze.pipeline.TailStage
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Future, Promise}
+import scala.collection.mutable.ArrayBuffer
+import org.http4s.blaze.util.Execution._
+
+import scala.util.{Failure, Success, Try}
+
+/** Combined [[WriteSerializer]] and [[ReadSerializer]] */
+trait Serializer[I] extends WriteSerializer[I] with ReadSerializer[I]
+
+/** Serializes write requests, storing intermediates in a queue */
+trait WriteSerializer[I] extends TailStage[I] { self =>
+
+  def maxWriteQueue: Int = 0
+
+  ////////////////////////////////////////////////////////////////////////
+
+  private val _serializerWriteLock = new AnyRef
+  private var _serializerWriteQueue = new ArrayBuffer[I]
+  private var _serializerWritePromise: Promise[Unit] = null
+
+  ///  channel writing bits //////////////////////////////////////////////
+  override def channelWrite(data: I): Future[Unit] = _serializerWriteLock.synchronized {
+    if (maxWriteQueue > 0 && _serializerWriteQueue.length > maxWriteQueue) {
+      Future.failed(new Exception(s"$name Stage max write queue exceeded: $maxWriteQueue"))
```

Comment:
How about an IllegalStateException?  Or even a Blaze-specific exception that extends it?

Do we want the exception to include the length and not just the cap?